mapiproxy/servers/default/emsmdb/emsmdbp_object.c

```diff
@@ -900,6 +900,95 @@ static int emsmdbp_copy_properties(struct emsmdbp_context *emsmdbp_ctx, struct e
+
+   \note This version is a copy of emsmdbp_copy_properties but which
+   removes the exclusion of PidTagChangeKey and
+   PidTagPredecessorChangeList properties.
+
+   \return MAPI_E_SUCCESS on success, otherwise MAPI error.
+ */
+static int emsmdbp_copy_properties_submit(struct emsmdbp_context *emsmdbp_ctx, struct emsmdbp_object *source_object, struct emsmdbp_object *dest_object, struct SPropTagArray *excluded_tags)
+{
+	TALLOC_CTX		*mem_ctx;
+	bool			properties_exclusion[65536];
+	struct SPropTagArray	*properties, *needed_properties;
+        void                    **data_pointers;
+        enum MAPISTATUS         *retvals = NULL;
+	struct SRow		*aRow;
+	struct SPropValue	newValue;
+	uint32_t		i;
+
+	mem_ctx = talloc_new(NULL);
+	OPENCHANGE_RETVAL_IF(!mem_ctx, MAPI_E_NOT_ENOUGH_MEMORY, NULL);
+
+	if (emsmdbp_object_get_available_properties(mem_ctx, emsmdbp_ctx, source_object, &properties) == MAPISTORE_ERROR) {
+		OC_DEBUG(0, "mapistore support not implemented yet - shouldn't occur\n");
+		talloc_free(mem_ctx);
+		return MAPI_E_NO_SUPPORT;
+	}
+
+	/* 1. Exclusions */
+	memset(properties_exclusion, 0, 65536 * sizeof(bool));
+
+	/* 1a. Explicit exclusions */
+	properties_exclusion[(uint16_t) (PidTagRowType >> 16)] = true;
+	properties_exclusion[(uint16_t) (PidTagInstanceKey >> 16)] = true;
+	properties_exclusion[(uint16_t) (PidTagInstanceNum >> 16)] = true;
+	properties_exclusion[(uint16_t) (PidTagInstID >> 16)] = true;
+	properties_exclusion[(uint16_t) (PidTagFolderId >> 16)] = true;
+	properties_exclusion[(uint16_t) (PidTagMid >> 16)] = true;
+	properties_exclusion[(uint16_t) (PidTagSourceKey >> 16)] = true;
+	properties_exclusion[(uint16_t) (PidTagParentSourceKey >> 16)] = true;
+	properties_exclusion[(uint16_t) (PidTagParentFolderId >> 16)] = true;
+	properties_exclusion[(uint16_t) (PidTagChangeNumber >> 16)] = true;
+
+	/* 1b. Request exclusions */
+	if (excluded_tags != NULL) {
+		for (i = 0; i < excluded_tags->cValues; i++) {
+			properties_exclusion[(uint16_t) (excluded_tags->aulPropTag[i] >> 16)] = true;
+		}
+	}
+
+	needed_properties = talloc_zero(mem_ctx, struct SPropTagArray);
+	needed_properties->aulPropTag = talloc_zero(needed_properties, void);
+	for (i = 0; i < properties->cValues; i++) {
+		if (!properties_exclusion[(uint16_t) (properties->aulPropTag[i] >> 16)]) {
+			SPropTagArray_add(mem_ctx, needed_properties, properties->aulPropTag[i]);
+		}
+	}
+
+	data_pointers = emsmdbp_object_get_properties(mem_ctx, emsmdbp_ctx, source_object, needed_properties, &retvals);
+	if (data_pointers) {
+		aRow = talloc_zero(mem_ctx, struct SRow);
+		for (i = 0; i < needed_properties->cValues; i++) {
+			if (retvals[i] == MAPI_E_SUCCESS) {
+				set_SPropValue_proptag(&newValue, needed_properties->aulPropTag[i], data_pointers[i]);
+				SRow_addprop(aRow, newValue);
+			}
+		}
+		if (emsmdbp_object_set_properties(emsmdbp_ctx, dest_object, aRow) != MAPISTORE_SUCCESS) {
+			talloc_free(mem_ctx);
```

Comment:
I'd be great to have a debug message with level 5 here as the returned value is the same in the next condition so it is impossible to retrace when  this error happens.